XCOM push ORA error code in OracleStoredProcedure

kolfild26 opened the pull request:
_**Summary.**_ ORA exit codes are controlled by the DB developer and could be a part of a dag logic. So in order to get access to these codes in dag, I propose an easy enhancement - push ORA exit code to XCOM when DatabaseError occurs.

One file changed - providers/oracle/operators/oracle.py

Motivation and detailed description - https://github.com/apache/airflow/discussions/23787 @potiuk

kolfild26 committed:
XCOM push ORA error code in OracleStoredProcedure

Motivation:
    https://github.com/apache/airflow/discussions/23787

ORA exit codes is controlled by the DB developer and could be a part of a dag logic.
So in order to get access to these codes in dag, I propose an easy enhancement - push ORA exit code to XCOM when DatabaseError occurs.

One file changed - providers/oracle/operators/oracle.py

uranusjr commented on airflow/providers/oracle/operators/oracle.py:
```diff
@@ -93,4 +95,10 @@ def __init__(
+        except oracledb.DatabaseError as e:
+            code, mesg = e.args[0].message[:-1].split(': ', 1)
+            ti.xcom_push(key='ORA', value=str(code.split('-')[1]))
+            raise AirflowException(f"Task failed with exit code: {code}")
```
Instead of coercing the exception to AirflowException, I would prefer this to simply re-raise the original exception:

```suggestion
            raise
```

kolfild26 replied:
@uranusjr  applied. Also removed import. thanks.

kolfild26 committed:
raise original instead of raise AirflowException

https://github.com/apache/airflow/pull/27319/files/daf4ddf632a6e9d2fcf58982a697c4d1e031364b#r1007791853

eladkal commented on airflow/providers/oracle/operators/oracle.py:
```diff
@@ -93,4 +94,10 @@ def __init__(
+            return hook.callproc(self.procedure, autocommit=True, parameters=self.parameters)
+        except oracledb.DatabaseError as e:
+            code, mesg = e.args[0].message[:-1].split(': ', 1)
+            ti.xcom_push(key='ORA', value=str(code.split('-')[1]))
```
Should we check for `self.do_xcom_push` ?

kolfild26 replied:
@eladkal  Do you mean checking whether `self.do_xcom_push` is `True` and only if so, then push to XCOM, right?

kolfild26 replied:
@eladkal I just tested, it works, no problem to introduce it, but as it's mentioned in https://github.com/apache/airflow/pull/27370 for many operators, pushing is not optional.
Again, if needed, will add.

eladkal replied:
The flag exist so this could be user choice. The fact that some operators dont use it is not something we should encourage.

kolfild26 replied:
@eladkal fixed, thanks.

bdsoha replied:
IMO, there is a difference between *data* returned from an operator and *meta-data* describing the operator's execution. The `self.do_xcom_push` controls the storing *(pushing)* of data returned by the logic of a given operator and not *meta-data*.

@eladkal What do you think?

eladkal replied:
Consider that not all users need this functionality. Why should we "spam" thier xcom table with data that has no value for them?
The use case of doing something in downstream task with error code of upstream task is not that common. at least from reviewing our issue tracker and questions on slack/other platforms.

kolfild26 replied:
@eladkal 

> Please add unit test to cover this change

Let me ask, the tests have to be placed in `airflow/tests/providers/oracle/operators/test_oracle.py` right?

eladkal replied:
yes

eladkal reviewed (commented):
Please add unit test to cover this change

kolfild26 committed:
Update oracle.py

kolfild26 committed:
added double quotes instaed of single quotes

bdsoha commented on airflow/providers/oracle/operators/oracle.py:
```diff
@@ -93,4 +94,11 @@ def __init__(
+            return hook.callproc(self.procedure, autocommit=True, parameters=self.parameters)
+        except oracledb.DatabaseError as e:
+            code, mesg = e.args[0].message[:-1].split(": ", 1)
+            if self.do_xcom_push is True:
```
No need to be so explicit

```suggestion
            if self.do_xcom_push:
```

kolfild26 committed:
simplify do_xcom_push check

Co-authored-by: Dov Benyomin Sohacheski <b@kloud.email>

kolfild26 committed:
re.search istead of split

kolfild26 committed:
handing ORA- not found in the begining of returned message

kolfild26 committed:
fixed mypy Handle Optionals

bdsoha commented on airflow/providers/oracle/operators/oracle.py:
```diff
@@ -98,7 +102,9 @@ def execute(self, context: Context):
+            if code_match is not None:
+                if self.do_xcom_push:
+                    if ti is not None:
+                        ti.xcom_push(key="ORA", value=code_match.group(1))
```
IMO, this is cognitively complex.

```suggestion
            if code_match and self.do_xcom_push and ti:
                ti.xcom_push(key="ORA", value=code_match.group(1))
```

kolfild26 committed:
simplify xcom_push ablility check

uranusjr commented on airflow/providers/oracle/operators/oracle.py:
```diff
@@ -93,4 +99,11 @@ def __init__(
+        except oracledb.DatabaseError as e:
+            code_match = re.search("^ORA-(\\d+):.+", str(e))
+            if code_match and self.do_xcom_push and ti:
+                ti.xcom_push(key="ORA", value=code_match.group(1))
```
```suggestion
            if not self.do_xcom_push or not ti:
                raise
            code_match = re.search("^ORA-(\\d+):.+", str(e))
            if code_match:
                ti.xcom_push(key="ORA", value=code_match.group(1))
```

This saves the regex match unless needed.

kolfild26 committed:
save the regex match unless needed

https://github.com/apache/airflow/pull/27319#discussion_r1015387980

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

uranusjr commented on airflow/providers/oracle/operators/oracle.py:
```diff
@@ -69,6 +72,9 @@ class OracleStoredProcedureOperator(BaseOperator):
     :param parameters: (optional, templated) the parameters provided in the call
+
+    :xcom ORA: Oracle db exit code. Optional (if self.do_xcom_push is True).
+        Numeric exit code that Oracle throws in case of failure.
```
```suggestion
    If *do_xcom_push* is *True*, the numeric exit code emitted by
    the database is pused to XCom under key ``ORA`` in case of failure.
```

uranusjr committed:
Docstring format

bdsoha commented on airflow/providers/oracle/operators/oracle.py:
```diff
@@ -69,6 +72,9 @@ class OracleStoredProcedureOperator(BaseOperator):
     :param parameters: (optional, templated) the parameters provided in the call
+
+    If *do_xcom_push* is *True*, the numeric exit code emitted by
+    the database is pused to XCom under key ``ORA`` in case of failure.
```
There is a spelling mistake causing the *build docs* step to fail.

```suggestion
    If *do_xcom_push* is *True*, the numeric exit code emitted by
    the database is pushed to XCom under key ``ORA`` in case of failure.
```

uranusjr replied:
I think the intended word is _pushed_.

kolfild26 committed:
spelling mistake fixed 'pused' to 'pushed'

Co-authored-by: Dov Benyomin Sohacheski <b@kloud.email>

kolfild26 committed:
Merge branch 'main' into main

kolfild26 committed:
test_push_oracle_exit_to_xcom added

kolfild26 committed:
fix execute method: context.get(task_instance) -> context[ti]

kolfild26 committed:
Merge branch 'main' into main
